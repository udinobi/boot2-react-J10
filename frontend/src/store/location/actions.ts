import { ActionCreator } from "redux"

import { Location } from "./types"

import { LocationChangedAction } from "./types"

export const changeCurrentLocation: ActionCreator<LocationChangedAction> = (location: Location) => ({
    payload: {
        location,
    },
    type: "location/LOCATION_CHANGED"
})
