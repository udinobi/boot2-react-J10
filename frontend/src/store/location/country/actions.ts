import axios, { AxiosResponse } from "axios"
import { ActionCreator, Dispatch } from "redux"

import {
    CountriesFailureAction, CountriesLoadingAction, CountriesRetrievedAction,
    Country, CountryActions, CountryActionType,
    CountrySelectedAction, CountryState
} from "./types"

// CountriesFailureAction --------------------------------------------------------------------------

const countriesFailure: ActionCreator<CountriesFailureAction> = () => ({
    type: CountryActionType.COUNTRIES_FAILURE
})

// CountriesLoadAction -----------------------------------------------------------------------------

const countriesLoading: ActionCreator<CountriesLoadingAction> = () => ({
    type: CountryActionType.COUNTRIES_LOADING
})

// Not checking yet, at this stage in our CSR app, if these properties are set or not.
// In a future SSR rendering feature branch we can also consider to implement logging
// to the server-side RT.
const url =
    `${process.env.REACT_APP_SERVICE_URL}${process.env.REACT_APP_LOAD_COUNTRIES_PATH}`

export const loadCountries = () =>
    async (dispatch: Dispatch<CountryActions>) => {
        dispatch(countriesLoading())

        try {
            const response: AxiosResponse<CountryState> = await axios.get(url)
            dispatch(countriesRetrieved(response.data))
        } catch (error) {
            dispatch(countriesFailure())
        }
    }
        
// CountriesRetrievedAction ------------------------------------------------------------------------

const countriesRetrieved: ActionCreator<CountriesRetrievedAction> = (countries: Country[]) => ({
    payload: { countries },
    type: CountryActionType.COUNTRIES_RETRIEVED,
})

// CountrySelectedAction ---------------------------------------------------------------------------

export const countrySelectedAction: ActionCreator<CountrySelectedAction> = (country: Country) => ({
    payload: { country },
    type: CountryActionType.COUNTRY_SELECTED
})

export const countrySelected = (country: Country) =>
    (dispatch: Dispatch<CountrySelectedAction>) => dispatch(countrySelectedAction(country))

// Action mapping ----------------------------------------------------------------------------------

export interface CountriesProps {
    countries: Country[]
    countrySelected: typeof countrySelected
    loadCountries: typeof loadCountries
}

export const mapCountryDispatchToProps = {
    countrySelected,
    loadCountries
}
    