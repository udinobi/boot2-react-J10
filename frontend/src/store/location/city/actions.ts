import { ActionCreator, Dispatch } from "redux"

import {
    CityActionType,
    Location, LocationSelectedAction,
    SuggestionsFailureAction, SuggestionsLookupAction,
    SuggestionsResetAction, SuggestionsRetrievedAction
} from "./types"

import { Country } from "../country/types"

// LocationSelectedAction --------------------------------------------------------------------------

const locationSelectedAction: ActionCreator<LocationSelectedAction> = (location: Location) => ({
    payload: { location },
    type: CityActionType.LOCATION_SELECTED
})

export const locationSelected = (location: Location) =>
    (dispatch: Dispatch<LocationSelectedAction | SuggestionsResetAction>) => {
        dispatch(suggestionsReset())
        dispatch(locationSelectedAction(location))
    }

// SuggestionsFailureAction ------------------------------------------------------------------------

export const suggestionsFailure: ActionCreator<SuggestionsFailureAction> = () => ({
    type: CityActionType.SUGGESTIONS_FAILURE
})

// SuggestionsLookupAction -------------------------------------------------------------------------

export const suggestionsLookup: ActionCreator<SuggestionsLookupAction> = (country: Country, locationTerm: string) => ({
    payload: { country, locationTerm },
    type: CityActionType.SUGGESTIONS_LOOKUP
})

// SuggestionsResetAction ------------------------------------------------------------------------

const suggestionsReset: ActionCreator<SuggestionsResetAction> = () => ({
    type: CityActionType.SUGGESTIONS_RESET
})

// SuggestionsRetrievedAction ----------------------------------------------------------------------

export const suggestionsRetrieved: ActionCreator<SuggestionsRetrievedAction> = (country: Country, suggestions: Location[]) => ({
    payload: { country, suggestions },
    type: CityActionType.SUGGESTIONS_RETRIEVED
})

// Action mapping ----------------------------------------------------------------------------------

export interface CityDispatchProps {
    locationSelected: typeof locationSelected
    suggestionsLookup: typeof suggestionsLookup
    suggestionsReset: typeof suggestionsReset
}

export const mapCityDispatchToProps = {
    locationSelected,
    suggestionsLookup,
    suggestionsReset
}
