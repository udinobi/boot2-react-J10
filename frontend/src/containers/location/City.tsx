import { AutoComplete } from "antd"
import React from "react"
import { connect } from "react-redux"

import styled from "styled-components"

import { option } from "ts-option"

import { initialCityState } from "../../store/location/city/reducer"

import { AppState } from "../../store/store"

import {
    CityProps,
    mapCityDispatchToProps
} from "../../store/location/city/actions"

import { CityState, Location } from "../../store/location/city/types"

export class City extends React.Component<CityProps, CityState> {

    private readonly autocomplete = styled(AutoComplete)`
        font-size: 1.1rem;
        width: 100%;
    `

    private readonly minLenSuggestionPrefix =
        +option(process.env.REACT_APP_MIN_LEN_SUGGESTION_PREFIX).getOrElse("3")

    private readonly Option = AutoComplete.Option

    constructor(props: CityProps) {
        super(props)

		this.state = initialCityState
        this.onSelect = this.onSelect.bind(this)
    }

    public componentWillReceiveProps(props: CityProps & CityState) {
        this.setState({
            country: props.country,
            suggestions: props.suggestions
        })
    }

    public render() {
        return (
            <this.autocomplete
                allowClear={true}
                defaultActiveFirstOption={false}
                onSearch={this.onSearch}
                onSelect={this.onSelect}
            >
                {this.state.suggestions.map(this.optionsFromSuggestions)}
            </this.autocomplete>
        )
    }

    private readonly optionsFromSuggestions = (suggestion: Location, index: number) =>
        <this.Option value={`${index}`} key={suggestion.name}>
            {`${suggestion.name} [${suggestion.coord.lat},${suggestion.coord.lon}]`}
        </this.Option>

    private readonly onSearch = (locationTerm: string) => {
        this.setState({ suggestions: [] })

        if (locationTerm.length < this.minLenSuggestionPrefix) {
            if (this.state.suggestions.length > 0) {
                this.props.suggestionsReset();
            }
        } else {
            this.state.country.map(country =>
                this.props.suggestionsLookup(country, locationTerm)
            )
        }
    }

    private readonly onSelect = (value: string) =>
        option(this.state.suggestions[+value])
            .map((suggestion: Location) => this.props.locationSelected(suggestion))
}

const mapStateToProps = (state: AppState): CityState => ({
    country: state.cityState.country,
    suggestions: state.cityState.suggestions
})

export default connect(mapStateToProps, mapCityDispatchToProps)(City);
