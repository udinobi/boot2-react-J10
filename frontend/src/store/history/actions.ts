import { ActionCreator, Dispatch } from "redux"

import { Location } from "../location/city/types"

import {
    HistoryActionType, HistoryItem, LocationReloadAction, LocationRemoveAction
} from './types'

// LocationReloadAction ----------------------------------------------------------------------------

const locationReloadAction: ActionCreator<LocationReloadAction> = (history: HistoryItem) => ({
    payload: { history },
    type: HistoryActionType.LOCATION_RELOAD
})

export const reloadLocation = (history: HistoryItem) =>
    (dispatch: Dispatch<LocationReloadAction>) => dispatch(locationReloadAction(history))

// LocationRemoveAction ----------------------------------------------------------------------------

const locationRemoveAction: ActionCreator<LocationRemoveAction> = (location: Location) => ({
    payload: { location },
    type: HistoryActionType.LOCATION_REMOVE
})

const removeLocation = (location: Location) =>
    (dispatch: Dispatch<LocationRemoveAction>) => dispatch(locationRemoveAction(location))

// Action mapping ----------------------------------------------------------------------------------

export interface HistoryDispatchProps {
    reloadLocation: typeof reloadLocation
    removeLocation: typeof removeLocation
}

export const mapHistoryDispatchToProps = {
    reloadLocation,
    removeLocation
}
