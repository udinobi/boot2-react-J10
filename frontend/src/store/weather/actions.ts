import { ActionCreator } from "redux"

import {
    WeatherActionType, WeatherData, WeatherDataFailureAction,
    WeatherDataRetrieveAction, WeatherDataRetrievedAction } from "./types"

import { Location } from "../location/city/types"

// WeatherDataFailureAction ------------------------------------------------------------------------

export const weatherDataFailure: ActionCreator<WeatherDataFailureAction> = () => ({
    type: WeatherActionType.WEATHER_DATA_FAILURE
})

// WeatherDataRetrieveAction -----------------------------------------------------------------------

export const weatherDataRetrieve: ActionCreator<WeatherDataRetrieveAction> =
    (location: Location, owmApiKey: string) => ({
        payload: { location, owmApiKey },
        type: WeatherActionType.WEATHER_DATA_RETRIEVE
    })

// WeatherDataRetrievedAction ----------------------------------------------------------------------

export const weatherDataRetrieved: ActionCreator<WeatherDataRetrievedAction> =
    (location: Location, weatherData: WeatherData) => ({
        payload: { location, weatherData },
        type: WeatherActionType.WEATHER_DATA_RETRIEVED
    })

// Action mapping ----------------------------------------------------------------------------------

export interface WeatherDispatchProps {
    weatherDataRetrieve: typeof weatherDataRetrieve
}

export const mapWeatherDispatchToProps = {
    weatherDataRetrieve
}
