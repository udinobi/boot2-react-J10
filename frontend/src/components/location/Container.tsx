import { Col, Row } from "antd"
import React from "react"
import ReactTooltip from 'react-tooltip'

import styled from "styled-components"

import City from "../../containers/location/City"
import Countries from "../../containers/location/Country"

import ReloadCountries from "./ReloadCountries"

const StyledCol = styled(Col)`
    @media (min-width: 1200px) {
      left: 4%;
    }
    @media (min-width: 1600px) {
      left: 8;
    }
    min-height: 180px;
    position: relative;
`

export default () => (
    <div>
        <Row>
            <div className="title-border">
                <h1 className="title">Location</h1>
            </div>
        </Row>
        <Row>
            <Col xs={24} sm={24} md={24} lg={24} xl={14} xxl={15} style={{ minHeight: 180 }}>
                <Row style={{ marginBottom: 6, textAlign: "center" }}>
                    <Col span={22}>
                        <label>Country</label>
                    </Col>
                </Row>
                <Row>
                    <Col xs={21} sm={21} md={21} lg={22}>
                        <Countries />
                    </Col>
                    {/* Even with hidden overflow, I was not able to use gutter for Row, so... rescue to inline styling */}
                    <Col span={1} />
                    <Col xs={2} sm={2} md={2} lg={1} style={{ marginTop: 2 }}>
                        <ReloadCountries />
                        <ReactTooltip id="ReloadCountries">
                            <div>Reload available countries</div>
                            <div>(because a new country was</div>
                            <div>&nbsp;probably added to the DB)</div>
                        </ReactTooltip>
                    </Col>
                </Row>
                <Row style={{ marginBottom: 6, marginTop: 28, textAlign: "center" }}>
                    <Col xs={21} sm={21} md={21} lg={22}>
                        <label>City</label>
                    </Col>
                </Row>
                <Row>
                    <Col xs={21} sm={21} md={21} lg={22}>
                        <City />
                    </Col>
                </Row>
            </Col>
            <Col xs={0} sm={0} md={0} lg={0} xl={1} xxl={2} />
            <StyledCol xs={0} sm={0} md={0} lg={0} xl={9} xxl={7}>
                <div style={{ fontSize : "1rem", position: "absolute", textAlign: "right", top: "60%", transform: "translateY(-50%)", width: "100%" }}>
                    Please, select here the location
                    you want the weather report for.
                </div>
            </StyledCol>
        </Row>
    </div>
)
