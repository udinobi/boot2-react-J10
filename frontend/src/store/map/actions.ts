import { reloadLocation } from '../history/actions'
import { locationSelected } from '../location/cityActions'

export interface MapDispatchProps {
    locationSelected: typeof locationSelected
    reloadLocation: typeof reloadLocation
}

export const mapMapDispatchToProps = {
    locationSelected,
    reloadLocation
}
