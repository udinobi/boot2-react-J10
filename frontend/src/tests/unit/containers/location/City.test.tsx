import React from "react"
import * as Renderer from 'react-test-renderer'

import  { City } from "../../../../containers/location/City"

import { CityProps } from "../../../../store/location/city/actions"

describe("--> Snapshot for Containers > location > 'City'", () => {
    let props: CityProps

    beforeEach(() => {
        props = {
            locationSelected: jest.fn(),
            suggestionsLookup: jest.fn(),
            suggestionsReset:  jest.fn()
        }
    })

    it("matches", () => {
        const component = Renderer.create(
            <City {...props} />    
        )

        const tree = component.toJSON()
        expect(tree).toMatchSnapshot();
    });
})
