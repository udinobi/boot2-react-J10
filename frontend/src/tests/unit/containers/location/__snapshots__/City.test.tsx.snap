// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`--> Snapshot for Containers > location > 'City' matches 1`] = `
<div
  className="sc-bdVaJa gikoeI ant-select-show-search ant-select-auto-complete ant-select ant-select-combobox ant-select-enabled ant-select-allow-clear"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={undefined}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={undefined}
  onMouseLeave={undefined}
  onTouchStart={[Function]}
  style={undefined}
>
  <div
    aria-autocomplete="list"
    aria-expanded={false}
    aria-haspopup="true"
    className="ant-select-selection
            ant-select-selection--single"
    role="combobox"
  >
    <div
      className="ant-select-selection__rendered"
    >
      <ul>
        <li
          className="ant-select-search ant-select-search--inline"
        >
          <div
            className="ant-select-search__field__wrap"
          >
            <input
              className="ant-input ant-select-search__field"
              disabled={false}
              onChange={[Function]}
              onKeyDown={[Function]}
              type="text"
              value=""
            />
            <span
              className="ant-select-search__field__mirror"
            >
              
              Â 
            </span>
          </div>
        </li>
      </ul>
    </div>
    <span
      className="ant-select-arrow"
      onClick={[Function]}
      style={
        Object {
          "WebkitUserSelect": "none",
          "userSelect": "none",
        }
      }
      unselectable="on"
    >
      <b />
    </span>
  </div>
</div>
`;
